File: app/src/main/java/com/dumuzeyn/mp3player/PlayerGradientBackground.java
package com.dumuzeyn.mp3player;

import android.graphics.Canvas;
import android.graphics.LinearGradient;
import android.graphics.Paint;
import android.graphics.Shader;
import android.os.SystemClock;
import android.view.View;
import android.content.Context;

final class PlayerGradientBackground extends View {
    interface Config {
        boolean animationsEnabled();

        boolean darkTheme();

        int baseColor();
    }

    private final Config config;
    private final Paint paint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final int startColor;
    private final int endColor;
    private boolean uiActive = true;

    PlayerGradientBackground(Context context, Config config, int startColor, int endColor) {
        super(context);
        this.config = config;
        this.startColor = startColor;
        this.endColor = endColor;
    }

    void setUiActive(boolean active) {
        uiActive = active;
        if (active) invalidate();
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        float phase = uiActive && config.animationsEnabled()
                ? (SystemClock.uptimeMillis() % 12000L) / 12000.0f
                : 0.18f;
        double angle = phase * Math.PI * 2.0;
        float centerX = getWidth() * 0.5f;
        float centerY = getHeight() * 0.5f;
        float radius = Math.max(getWidth(), getHeight()) * 0.72f;
        float dx = (float) Math.cos(angle) * radius;
        float dy = (float) Math.sin(angle) * radius;
        int baseColor = config.baseColor();
        int purpleTint = blend(baseColor, this.startColor,
                config.darkTheme() ? 0.82f : 0.38f);
        int yellowTint = blend(baseColor, this.endColor,
                config.darkTheme() ? 0.72f : 0.30f);
        paint.setShader(new LinearGradient(centerX - dx, centerY - dy, centerX + dx, centerY + dy,
                new int[]{baseColor, purpleTint, yellowTint, baseColor},
                new float[]{0.0f, 0.38f, 0.68f, 1.0f}, Shader.TileMode.CLAMP));
        canvas.drawRect(0.0f, 0.0f, getWidth(), getHeight(), paint);
        if (uiActive && config.animationsEnabled() && isAttachedToWindow()) {
            postInvalidateDelayed(40L);
        }
    }

    private static int blend(int first, int second, float amount) {
        float inverse = 1.0f - amount;
        int red = Math.round(((first >> 16) & 255) * inverse + ((second >> 16) & 255) * amount);
        int green = Math.round(((first >> 8) & 255) * inverse + ((second >> 8) & 255) * amount);
        int blue = Math.round((first & 255) * inverse + (second & 255) * amount);
        return 0xff000000 | (red << 16) | (green << 8) | blue;
    }
}

File: app/src/main/java/com/dumuzeyn/mp3player/PlayerGradientBackground.kt
package com.dumuzeyn.mp3player

import android.content.Context
import android.graphics.Canvas
import android.graphics.LinearGradient
import android.graphics.Paint
import android.graphics.Shader
import android.os.SystemClock
import android.view.View
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.roundToInt
import kotlin.math.sin

class PlayerGradientBackground(
    context: Context,
    private val config: Config,
    private val startColor: Int,
    private val endColor: Int,
) : View(context) {
    interface Config {
        fun animationsEnabled(): Boolean
        fun darkTheme(): Boolean
        fun baseColor(): Int
    }

    private val paint = Paint(Paint.ANTI_ALIAS_FLAG)
    private var uiActive = true

    fun setUiActive(active: Boolean) {
        uiActive = active
        if (active) invalidate()
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        val phase = if (uiActive && config.animationsEnabled()) {
            (SystemClock.uptimeMillis() % ANIMATION_DURATION_MS) / ANIMATION_DURATION_MS.toFloat()
        } else {
            STATIC_PHASE
        }
        val angle = phase * Math.PI * 2.0
        val centerX = width * 0.5f
        val centerY = height * 0.5f
        val radius = max(width, height) * 0.72f
        val dx = cos(angle).toFloat() * radius
        val dy = sin(angle).toFloat() * radius
        val baseColor = config.baseColor()
        val purpleTint = blend(baseColor, startColor, if (config.darkTheme()) 0.82f else 0.38f)
        val yellowTint = blend(baseColor, endColor, if (config.darkTheme()) 0.72f else 0.30f)
        paint.shader = LinearGradient(
            centerX - dx,
            centerY - dy,
            centerX + dx,
            centerY + dy,
            intArrayOf(baseColor, purpleTint, yellowTint, baseColor),
            floatArrayOf(0f, 0.38f, 0.68f, 1f),
            Shader.TileMode.CLAMP,
        )
        canvas.drawRect(0f, 0f, width.toFloat(), height.toFloat(), paint)
        if (uiActive && config.animationsEnabled() && isAttachedToWindow) {
            postInvalidateDelayed(FRAME_DELAY_MS)
        }
    }

    private fun blend(first: Int, second: Int, amount: Float): Int {
        val inverse = 1f - amount
        val red = (((first shr 16) and 255) * inverse + ((second shr 16) and 255) * amount).roundToInt()
        val green = (((first shr 8) and 255) * inverse + ((second shr 8) and 255) * amount).roundToInt()
        val blue = ((first and 255) * inverse + (second and 255) * amount).roundToInt()
        return -0x1000000 or (red shl 16) or (green shl 8) or blue
    }

    private companion object {
        const val ANIMATION_DURATION_MS = 12_000L
        const val FRAME_DELAY_MS = 40L
        const val STATIC_PHASE = 0.18f
    }
}
